proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,161 @@
+  In *1 requests flow `DOWNSTREAM` which makes no sense.
+
+### General short-comings/weaknesses
+
+* The metrics endpoint populates `HELP` for each metric, but this is populated with nothing more than the metric's name.
+  We ought to be giving a clear description of the metric.
+* Some of the label names don't follow Prometheus naming conventions.  Prometheus recommends snake case whereas we have used camel case in some cases.
+
+## Use Cases
+
+In general, metrics exist to help a user understand the health and performance of the proxy. They help inform
+capacity planning, drive system alerts reporting abnormal conditions and help with problem-solving (things like
+diagnosing failures or system slow-downs).
+
+Introducing a proxy to a Kafka system will increase latency of the system. Users will need a mechanism to understand how
+much latency is being added for their particular use-cases and have the ability to monitor this latency over time.
+
+Kroxylicious Filters themselves have the potential to alter behaviour of the Kafka system.  Users will need insights into
+performance and other issues that might result from filter behaviour. Specifically:
+
+* filters have the ability to change the size of the RPCs as they traverse the proxy (record encryption)
+* filters have the ability to change the flow of messages within the Kafka system. Specifically, they can short-circuit
+  requests by producing their own responses (the broker never sees the request) and introduce their own requests.
+
+## Proposal
+
+### Deprecate `kroxylicious_inbound*`/`kroxylicious_payload_size_bytes` replace with new `kroxylicious_(downstream|upstream)_kafka_rpc_message_size_bytes` metrics
+
+This proposal suggests that we deprecate the `kroxylicious_inbound*`/`kroxylicious_payload_size_bytes` metrics and replace
+with two new **distribution** (aka histogram) metrics `kroxylicious_downstream_kafka_rpc_message_size_bytes` and `kroxylicious_upstream_kafka_rpc_message_size_bytes`.
+
+The new "message_size" differ from the old "payload_size" in several key ways:
+1. New "message_size" will be used to track both **opaque** and **decoded** messages ("payload_size" tracked only decoded messages).
+2. Downstream "message size" will track [message size](https://kafka.apache.org/protocol#:~:text=DESCRIPTION-,message_size,-The%20message_size%20field)
+   as messages come in from or leave for the client.  Upstream "message size" will track message size as messages go to or arrive from the broker.
+3. The design of the metric allows it to be used to track messages that originate within a filter.
+4. The design of metric allow a message-size histogram to be added.  This would allow the collection of statistical information
+   on, say, the size of the produce request or the fetch response.  If added, we would make this configurable.
+
+In Micrometer, `Distribution` type metrics actually comprise three Prometheus metrics `_sum`, `_count` and `_max`.  This means that
```

Comment:
I'm in two minds about this.  On the one hand, this approach saves an additional counter.  On the other hand, it feels wrong to have the user rely on something with the words `message_size` `kroxylicious_downstream_kafka_rpc_message_size_bytes_count`.   It is counter intuitive.  I'd prefer an explicit counter.

Reply:
@SamBarker any opinion on this?  I'm veering towards adding counter metrics kroxylicious_(up|down)stream_kafka_message_count which the same set of labels as I think clean API wins rather than relying on the fact that distribution (histogram) metric actually comprises a _count metric too.

Reply:
Urgh it's a hard one. As you say the metrics are public API, and consistency and clarity are important.

In isolation I'd definitely say we should have a separate counter but we are in the context of Prometheus metrics with their own conventions. 

I'm not sure that it would be helpful to the end users to have duplicate counters...

🤷

Reply:
I think having another counter would be better to avoid confusions since `message_size` means a completely different thing then `message_count`